### ISSUE TITLE
Add DCT ops in fftpack

### ISSUE THREAD

**Issue by brianwa84** [google/jax#1839]
Reference implementations in terms of FFT may be found in TF:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/signal/dct_ops.py#L53

**Comment by shoyer** [comment]
We would probably need `rfft` and `irfft` first. I made a little bit of progress in https://github.com/google/jax/pull/1657, but got stuck on the transpose rules. It looks those are indeed somewhat non-trivial:
https://github.com/tensorflow/tensorflow/blob/r2.0/tensorflow/python/ops/signal/fft_ops.py#L218

**Comment by brianwa84** [comment]
If you assume an even-sized dimension, here are impls for i/rfft:

def rfft(x): x = tf.convert_to_tensor(x, dtype=tf.float32) return
fft(tf.cast(x, tf.complex64))[..., :tf.shape(x)[-1]//2+1] def
irfft(x,n=None): x = tf.concat([x, tf.math.conj(x[..., ::-1][..., 1:-1])],
axis=-1) return tf.math.real(ifft(x))
Brian Patton | Software Engineer | bjp@google.com

**Comment by shoyer** [comment]
I was thinking that there is no point in adding rfft without the efficiency gain, but it's a fair point that some users might just want them for the API convenience.

**Comment by joglekara** [comment]
Hmm, do we have any further thoughts on this? 

Should we target an efficient rfft implementation or just implement a `np.real()` call around the complex `fft` routines?

Would we be adding significant inertia against working on the former by implementing the latter?

**Comment by shoyer** [comment]
I think it would be totally reasonable to start with the easy wrapper approach based on `real(fft(x))`.

Conceivably we could even still use the XLA `rfft` functions for cases where we don't need a `transpose` (we could write only the `transpose` rule in terms of the complex FFTs)

**Comment by brianwa84** [comment]
What is the transpose you're talking about? Why not always do rfft and
irfft in xla directly?

**Comment by brianwa84** [comment]
Xla fft takes an fft_type which can be any of fft, ifft, rfft, or irfft.

**Comment by shoyer** [comment]
To implement backwards gradientd in JAX, we need a JVP rule and a transpose rule.

The JVP rule is easy (because FFTs are already linear) but the transpose rule (the same as the gradient definition in TensorFlow) for rfft can't be defined in terms of the other FFTs.

**Comment by brianwa84** [comment]
I see, so we would need to port over this helper:
https://github.com/sourcecode369/tensorflow-1/blob/63f319f30302304506e1a535856a68106f396524/tensorflow/python/ops/signal/fft_ops.py#L233

**Comment by shoyer** [comment]
I got transpose rules for rfftn and irfftn working over in https://github.com/google/jax/pull/1657, which is now ready for review!

**Comment by nmheim** [comment]
bump! :)

**Comment by hawkinsp** [comment]
@juliuskunze Neat! Are you interested in perhaps sending a PR adding them to JAX, perhaps under the `jax.scipy` namespace?

**Comment by juliuskunze** [comment]
@hawkinsp https://github.com/google/jax/pull/7617
